### PR TITLE
feat(ethernet): Added Ethernet support

### LINKED ISSUE #48: Ethernet connection crashing
Hi Lennart!

I'm having troubles trying to start a Telnet Server over a wired Ethernet connection (W5100 shield). Already checked #22 and #27 but I can't find my problem related to them.

I'm passing `false` as second argument to `telnet.begin()` and it's avoiding the WiFi check as expected, but later the code throws:

```
assert failed: tcpip_send_msg_wait_sem IDF/components/lwip/lwip/src/api/tcpip.c:455 (Invalid mbox)

Backtrace: 0x40084109:0x3ffb1e90 0x4008e535:0x3ffb1eb0 0x40094161:0x3ffb1ed0 0x4011bb55:0x3ffb2000 0x4012c6b5:0x3ffb2030 0x4012c715:0x3ffb2050 0x4011b2f1:0x3ffb20a0 0x400e9576:0x3ffb20c0 0x400e95f4:0x3ffb2110 0x400f9917:0x3ffb2130 0x400e02fb:0x3ffb2170 0x400de137:0x3ffb21a0 0x400dee07:0x3ffb21c0 0x400dfeb2:0x3ffb2240 0x400e0f54:0x3ffb2270 0x40101751:0x3ffb2290

  #0  0x40084109:0x3ffb1e90 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:408
  #1  0x4008e535:0x3ffb1eb0 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:137
  #2  0x40094161:0x3ffb1ed0 in __assert_func at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/assert.c:85
  #3  0x4011bb55:0x3ffb2000 in tcpip_send_msg_wait_sem at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/lwip/lwip/src/api/tcpip.c:455 (discriminator 1)
  #4  0x4012c6b5:0x3ffb2030 in netconn_apimsg at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/lwip/lwip/src/api/api_lib.c:136
  #5  0x4012c715:0x3ffb2050 in netconn_new_with_proto_and_callback at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/lwip/lwip/src/api/api_lib.c:166
  #6  0x4011b2f1:0x3ffb20a0 in lwip_socket at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/lwip/lwip/src/api/sockets.c:1774
  #7  0x400e9576:0x3ffb20c0 in WiFiServer::begin(unsigned short, int) at /Users/alberto/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32/include/lwip/lwip/src/include/lwip/sockets.h:656
      (inlined by) WiFiServer::begin(unsigned short, int) at /Users/alberto/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiServer.cpp:80
  #8  0x400e95f4:0x3ffb2110 in WiFiServer::begin(unsigned short) at /Users/alberto/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiServer.cpp:70
  #9  0x400f9917:0x3ffb2130 in ESPTelnetBase::begin(unsigned short, bool) at .pio/libdeps/esp32dev/ESP Telnet/src/ESPTelnetBase.cpp:21
  #10 0x400e02fb:0x3ffb2170 in vserial_listen(int, int (*)(char*), bool) at src/VSerial.cpp:256
  #11 0x400de137:0x3ffb21a0 in InitVTerm() at src/Trh32_Term.cpp:330
  #12 0x400dee07:0x3ffb21c0 in Trh32_Term_tick() at src/Trh32_Term.cpp:394
  #13 0x400dfeb2:0x3ffb2240 in profile(char const*, unsigned long, void (*)()) at src/UTILS.cpp:534
  #14 0x400e0f54:0x3ffb2270 in loop() at src/main.cpp:128
  #15 0x40101751:0x3ffb2290 in loopTask(void*) at /Users/alberto/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:50

ELF file SHA256: 3ecc7d947ae65388

Rebooting...
```
The stack trace shows:
Screen Shot 2023-07-17 at 20 09 54

Using the  debugger I could follow the offending line to the `WiFiServer::begin` method:
```
void WiFiServer::begin(uint16_t port, int enable){
  if(_listening)
    return;
  if(port){
      _port = port;
  }
  struct sockaddr_in server;
  sockfd = socket(AF_INET , SOCK_STREAM, 0);                        // <<<=== Panic is thrown HERE
  if (sockfd < 0)
    return;
  setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int));
  server.sin_family = AF_INET;
  server.sin_addr.s_addr = _addr;
  server.sin_port = htons(_port);
  if(bind(sockfd, (struct sockaddr *)&server, sizeof(server)) < 0)
    return;
  if(listen(sockfd , _max_clients) < 0)
    return;
  fcntl(sockfd, F_SETFL, O_NONBLOCK);
  _listening = true;
  _noDelay = false;
  _accepted_sockfd = -1;
```

The full code is huge but the library initialization portion is something like:
```
      static byte_8  EthMAC[6];

     // MAC init code here...

      Ethernet.init(ETH_CS);            // ETH_CS: chip select
      Ethernet.begin(EthMAC);

     telnet.onConnect(onTelnetConnect);
     telnet.setLineMode(true);
  
     telnet.begin(port, false);
     ...
```
Due to some related comment read somewhere I also tried pre-initializing Esp Wifi code with the following code with no luck:
```
const wifi_init_config_t wconfig = WIFI_INIT_CONFIG_DEFAULT();
esp_wifi_init(&wconfig);  
```

The telnet server is working fine on Wifi either in AP or STA modes.

Any hint would be greatly appreciated!

Alberto

### PR DESCRIPTION
**Pull request by AlbertoLopSie**
# Abstracted TCPClient and TCPServer classes to make as transparent as possible to switch between network interfaces

Hi Lennart!

I managed to create a transparency class to abstract as much as possible the differences between WiFi and Ethernet interfaces in your Telnet library.

I feel a little rusted on C++ inheritance and probably this same code could be enhanced and made clearer, but at least it:
 - Does not break existing WiFi code
 - It modifies very liitle ESPTelnet code and concentrates on Client and Server classes, so I don't think it could impact performance on WiFi users (the only ones that may exist by now)
 - It works!

Please fell free to add or modify anything you like  (It's YOUR library after all  :-) ) and merge it or not to the main branch. I will use the fork in my repo for my project because I'm in a hurry and it fits my needs. When, in the future, you feel satisfied with the tests you consider and finally merge it, future versions of my project will link it.

Cheers!!

### CODE REVIEW

File: src/ESPTelnetBase.cpp [hunk 1]
```diff
@@ -2,22 +2,24 @@
...
-bool ESPTelnetBase::begin(uint16_t port /* = 23 */, bool checkConnection /* = true */) {
+bool ESPTelnetBase::begin(uint16_t port /* = 23 */, bool checkWiFiConnection /* = true */, bool useEthernet /* = false */)
+{
   ip = "";
-  if (checkConnection) {
+  if (checkWiFiConnection) {
     // connected to WiFi or is ESP in AP mode?
     if (WiFi.status() != WL_CONNECTED && !_isIPSet(WiFi.softAPIP())) return false;
   }
   server_port = port;
```
Comment (LennartHennigs) [inline src/ESPTelnetBase.cpp:21]: storing the type - needed for is connected

File: src/ESPTelnetBase.h [hunk 2]
```diff
@@ -21,24 +21,112 @@
...
 
 /////////////////////////////////////////////////////////////////
 
-using TCPClient = WiFiClient;
-using TCPServer = WiFiServer;
+class TCPClient : public WiFiClient, EthernetClient {
+  public:
+    TCPClient() : _useEthernet(false) {}
+
+// Convenience methods
```
Comment (LennartHennigs) [inline src/ESPTelnetBase.h:36]: the parameter one only exists for the ESP32

File: src/ESPTelnetBase.h [hunk 3]
```diff
@@ -21,24 +21,112 @@
...
+    bool useEthernet(void) { return _useEthernet; }
+    const char *connection(void) { return _useEthernet ? "Ethernet" : "WiFi"; }
+
+// Methods override
+    int read() { return _useEthernet ? EthernetClient::read() : WiFiClient::read(); }
+    int read(uint8_t *buf, size_t size) { return _useEthernet ? EthernetClient::read(buf, size) : WiFiClient::read(buf, size); }
+    size_t write(uint8_t data) { return _useEthernet ? EthernetClient::write(data) : WiFiClient::write(data); }
+    size_t write(const uint8_t *buf, size_t size) { return _useEthernet ? EthernetClient::write(buf, size) : WiFiClient::write(buf, size); }
+    int available() { return _useEthernet ? EthernetClient::available() : WiFiClient::available(); }
+    int peek() { return _useEthernet ? EthernetClient::peek() : WiFiClient::peek(); }
```
Comment (LennartHennigs) [inline src/ESPTelnetBase.h:46]: needed a status() func., too

File: src/ESPTelnetBase.h [hunk 4]
```diff
@@ -21,24 +21,112 @@
...
+    size_t print(const char s[])                  { return _useEthernet ? EthernetClient::print(s)  :  WiFiClient::print(s); }
+    size_t print(char c)                          { return _useEthernet ? EthernetClient::print(c)  :  WiFiClient::print(c); }
+    size_t print(unsigned char c, int b = DEC)    { return _useEthernet ? EthernetClient::print(c, b)  :  WiFiClient::print(c, b); }
+    size_t print(int c, int b = DEC)              { return _useEthernet ? EthernetClient::print(c, b)  :  WiFiClient::print(c, b); }
+    size_t print(unsigned int c, int b = DEC)     { return _useEthernet ? EthernetClient::print(c, b)  :  WiFiClient::print(c, b); }
+    size_t print(long l, int b= DEC)              { return _useEthernet ? EthernetClient::print(l, b)  :  WiFiClient::print(l, b); }
+    size_t print(unsigned long l, int b = DEC)    { return _useEthernet ? EthernetClient::print(l, b)  :  WiFiClient::print(l, b); }
+    size_t print(long long l, int b = DEC)        { return _useEthernet ? EthernetClient::print(l, b)  :  WiFiClient::print(l, b); }
+    size_t print(unsigned long long l, int b = DEC) { return _useEthernet ? EthernetClient::print(l, b)  :  WiFiClient::print(l, b); }
```
Comment (LennartHennigs) [inline src/ESPTelnetBase.h:74]: threw conversation errors

File: src/ESPTelnetBase.h [hunk 5]
```diff
@@ -21,24 +21,112 @@
...
+    size_t println(char c)                        { return _useEthernet ? EthernetClient::println(c)  :  WiFiClient::println(c); }
+    size_t println(unsigned char c, int b = DEC)  { return _useEthernet ? EthernetClient::println(c, b)  :  WiFiClient::println(c, b); }
+    size_t println(int c, int b = DEC)            { return _useEthernet ? EthernetClient::println(c, b)  :  WiFiClient::println(c, b); }
+    size_t println(unsigned int c, int b = DEC)   { return _useEthernet ? EthernetClient::println(c, b)  :  WiFiClient::println(c, b); }
+    size_t println(long l, int b = DEC)           { return _useEthernet ? EthernetClient::println(l, b)  :  WiFiClient::println(l, b); }
+    size_t println(unsigned long l, int b = DEC)  { return _useEthernet ? EthernetClient::println(l, b)  :  WiFiClient::println(l, b); }
+    size_t println(long long l, int b = DEC)      { return _useEthernet ? EthernetClient::println(l, b)  :  WiFiClient::println(l, b); }
+    size_t println(unsigned long long l, int b = DEC) { return _useEthernet ? EthernetClient::println(l, b)  :  WiFiClient::println(l, b); }
```
Comment (LennartHennigs) [inline src/ESPTelnetBase.h:89]: threw conversation error

**Comment by LennartHennigs** [comment]
I made some changes to your code – now it compiles on ESP32 and ESP8266 on both Arduino IDE versions (on Mac).

I moved it to this branch to test it a bit further:
https://github.com/LennartHennigs/ESPTelnet/tree/pr-ethernet-static

**Comment by AlbertoLopSie** [comment]
Hi Lennart,

I noted you merged this PR to the `pr-ethernet-static` but no the main branch. This has over 50 additional commits leaving the Ethernet supporting branch old.

Are you handling Ethernet connections on a different way on the main branch?

**Comment by LennartHennigs** [comment]
Note for the record: this PR was an early proof-of-concept for SPI Ethernet support (merged only to the `pr-ethernet-static` spike branch, never to `main`). The approach — multiple-inheriting both `WiFiClient`/`WiFiServer` and `EthernetClient`/`EthernetServer` behind a runtime `_useEthernet` flag, plus an unconditional `#include <Ethernet.h>` — was ultimately not viable for `main`: it forces the Ethernet library on all WiFi-only users, doubles per-client RAM, and introduces a diamond-inheritance ambiguity.

The library instead went with native RMII Ethernet via `ETH.h` (shared lwIP stack, no extra dependency). Full context and the final decision are in #48, which is now closed. Leaving this here for history — thanks again @AlbertoLopSie for the exploration.

**Pull request merged**